### PR TITLE
AWS app access Timestream Endpoint mode and JDBC support

### LINKED ISSUE #15574: AWS app to cache AWS session credentials
What would you like Teleport to do?
Teleport Application Service currently make a new AssumeRole call to get AWS credentials for every incoming AWS API request.

It would be great the AWS credentials can be cached to avoid an STS call to AWS every time. Especially with the support of `tsh proxy aws`, users can run tools like `terraform` that possibly make hundreds of AWS calls in the same session.

What problem does this solve?
Avoid spamming AWS STS service.

### PR DESCRIPTION
**Pull request by greedy52**
This change fixes Timestream in Endpoint mode. 
```
$ tsh aws --endpoint-url timestream-query list-scheduled-queries --region us-west-2

An error occurred (UnknownOperationException) when calling the ListScheduledQueries operation: 
ERROR: exit status 254
```
Also, this change enables Timestream JDBC support as the Timestream JDBC drivers does NOT support HTTPS proxy so `--endpoint-url` must be used.

Also fixes #15574

### Setup a Timestream database
https://github.com/awslabs/amazon-timestream-tools/tree/mainline/sample_apps/python
Note that not all regions support Timestream yet.

### Setup Teleport
Make sure the server is up-to-date (see #21766).

[optional] pre-populate defaults for `tsh` template values:
```
export TELEPORT_AWS_ACCESS_KEY_ID="example-access-key-id"
export TELEPORT_AWS_SECRET_ACCESS_KEY="example-secret-access-key"
export TELEPORT_AWS_REGION="us-west-2"
export TELEPORT_AWS_KEYSTORE=$(/usr/libexec/java_home)/lib/security/cacerts
```
Then login the AWS app. [optional] You can run `list-tables` to verify the table is there.
```
$ tsh login ....
$ tsh apps login aws
$ tsh aws timestream-write list-tables --region us-west-2
```
Then launch the local proxy
```
$ tsh proxy aws -p8888 -f timestream-jdbc --endpoint-url
```

Follow the instruction to add the CA to the keystore.

### Setup JDBC client
I personally find that [SquirrelSQL](https://squirrel-sql.sourceforge.io/) is pretty good on MacOS.

(If you have the official java 8.dmg installed on MacOS, SquirreSQL will not run as java 8 is old. Uninstall it and install a newer openjdk) 

First, add a driver in SquirrelSQL, the jar can be found [here](https://docs.aws.amazon.com/timestream/latest/developerguide/JDBC.configuring.html) (note, use the "shaded" jar)
driver

Then, add an "alias" connection. You can just copy/paste the connection URL from `tsh proxy aws` command.
alias

Then connect. That's it!
connect

### CODE REVIEW

File: lib/srv/app/aws/endpoints.go
Comment (Tener) [inline lib/srv/app/aws/endpoints.go:250]: The duplication between `resolveTimestreamQueryEndpoint` and `resolveTimestreamWriteEndpoint` is unfortunate. Can we share this logic somehow?

File: lib/srv/app/aws/endpoints.go
Comment (Tener) [inline lib/srv/app/aws/endpoints.go:346]: Comments would be welcome.

File: lib/srv/app/aws/endpoints.go
Comment (Tener) [inline lib/srv/app/aws/endpoints.go:76]: A link to the docs on legacy mode would be welcome.

File: lib/srv/app/aws/handler.go
Comment (Tener) [inline lib/srv/app/aws/handler.go:95]: Can we hide `SigningService` and always initialize it here? This would ensure consistency in how it is initialized.

File: lib/utils/aws/credentials.go
Comment (Tener) [inline lib/utils/aws/credentials.go:94]: The ~13 minutes TTL feels like a lot. There shouldn't be any performance problems if we were to lower this to something like 60 seconds?

Comment (greedy52) [reply lib/utils/aws/credentials.go:94]: sounds good. Will do one minute as other places. I don't think there will be a performance problem. We are currently doing this for every single API call. Even caching for a second will save a lot when using things like terraform.

**Review by GavinFrazar — COMMENTED** [review]
mostly looks good, I like the refactoring for signing service and credential fetching 😄 but I have one concern around endpoint discovery to discuss before I approve

File: lib/srv/app/aws/endpoints.go
Comment (greedy52) [reply lib/srv/app/aws/endpoints.go:201]: We only cache it for a minute in `FnCache` which is the minimum of `CachePeriodInMinutes` (tho it seems `CachePeriodInMinutes` is usually 1440m (24h) by default). Note that since the endpoint can be different depending on who calls, I am using the user sessions' credentials here as part of the cache key. I can add a comment on the TTL here.

Seems a lot of work is required to get `crr.EndpointCache` working. And it only supports `string` key, and the `crr.Endpoint` is not thread-safe. So personally I would prefer to keep it simple with `FnCache`.

Comment (mdwn) [reply lib/srv/app/aws/endpoints.go:203]: I'm +1 to using AWS SDK's endpoint cache if we can, acknowledging the issue with thread safety. The `Discoverer` looks pretty minor to implement -- does the thread safety of `EndpointCache` matter here given that the Cache itself seems to be thread safe?

**Review by mdwn — APPROVED** [review]
I don't have anything much to add here. Save addressing the existing issues, LGTM.

File: lib/srv/app/aws/endpoints.go
Comment (greedy52) [reply lib/srv/app/aws/endpoints.go:203]: I haven't tried an example with race detector. so this is speculation by inspecting this code
https://github.com/aws/aws-sdk-go/blob/5fabc200455abcde45b34541ee557eb910ab0bd6/aws/crr/cache.go#L28-L42

The map is a sync.Map but `ev.Prune()` is not concurrent safe. But I guess two concurrent `ev.Prune` probably end up the same thing.

I don't mind giving it a try or maybe implementing our own. but let's say the API returns 24h. Do we want to cache it for that long?

Comment (GavinFrazar) [reply lib/srv/app/aws/endpoints.go:203]: regarding thread safety: I think it should be fine because the endpoint cache is storing/loading endpoint structs by value, so the endpoint returned from the sync.Map load will be a copy.

timestreamquery has an implementation of discover here which you could pretty much copy https://github.com/aws/aws-sdk-go/blob/5fabc200455abcde45b34541ee557eb910ab0bd6/service/timestreamquery/api.go#L493-L538

>  Note that since the endpoint can be different depending on who calls

I understand that this would be the case for different AWS accounts, but is it true that the endpoint can differ depending on caller within the same account?

> but let's say the API returns 24h. Do we want to cache it for that long?

The endpoint cache is an LRU cache so you can control the size of it. Looks like AWS opted to just use 10 for the size in their usage.

My main concern was that we might use an endpoint after it had expired. If the minimum TTL is 1 minute then that isn't an issue, although we will probably make a lot more endpoint discovery api calls than necessary. I leave it up to you to decide, so I won't block the PR anymore

**Review by GavinFrazar — APPROVED** [review]

File: lib/srv/app/aws/endpoints.go
Comment (GavinFrazar) [reply lib/srv/app/aws/endpoints.go:203]: > We do support multiple accounts through AssumeRole. I guess we can extract Account ID from the role ARN if we want.

Oh, right 👍 

> Seems the endpoint cache is not LRU but it deletes a random entry.

You're right I think, the godoc for it seems to be inaccurate then.

Not sure what to pick for a cache size, but you can limit the number of endpoint addresses you add to each `crr.Endpoint` (as the code is doing now by just picking the first endpoint address)

**Pull request closed without merging**
